[lldb] Update the String table offset based on the DWARF format

HemangGadhavi opened the pull request:
This PR is updating the string table offset (DW_AT_str_offsets_base which is introduced in `DWARF5`) based on the DWARF format, as per the DWARF specification; For the 32-bit DWARF format, each offset is 4 bytes long; for the 64-bit DWARF format, each offset is 8 bytes long.

HemangGadhavi committed:
[lldb] Updated the String table offset based on the DWARF format

HemangGadhavi commented:
@labath @DavidSpickett 
Related to PR https://github.com/llvm/llvm-project/issues/135208 
Please have review once.  DW_AT_str_offsets_base was missing to handle for DWARF64.

labath reviewed (commented):
Could use a test case.

DavidSpickett commented:
This is the relevant part of the standard, correct?

```
7.5.5 Classes and Forms

18 • stroffsetsptr
19 This is an offset into the .debug_str_offsets section
20 (DW_FORM_sec_offset). It consists of an offset from the beginning of the
21 .debug_str_offsets section to the beginning of the string offsets
22 information for the referencing entity. It is relocatable in a relocatable object
23 file, and relocated in an executable or shared object file. In the 32-bit
24 DWARF format, this offset is a 4-byte unsigned value; in the 64-bit DWARF
25 format, it is an 8-byte unsigned value (see Section 7.4 on page 196).
```
(https://dwarfstd.org/doc/DWARF5.pdf)

I see also "7.4 32-Bit and 64-Bit DWARF Formats", is there anything else in there we need to handle? (I don't know enough to guess the effects of getting those wrong)

I'm also going to run the test suite and see what fails. The baseline is dwarf32:
```
LLDB_TEST_USER_ARGS              -E;\"-gdwarf32\"

********************
Failed Tests (5):
  lldb-api :: commands/frame/language/TestGuessLanguage.py
  lldb-api :: functionalities/multiple-slides/TestMultipleSlides.py
  lldb-api :: functionalities/step-avoids-no-debug/TestStepNoDebug.py
  lldb-api :: lang/c/conflicting-symbol/TestConflictingSymbol.py
  lldb-api :: lang/cpp/incomplete-types/members/TestCppIncompleteTypeMembers.py
```
All of these manipulate debug info flags, so this is expected.

Now DWARF64 current main branch:
```
LLDB_TEST_USER_ARGS              -E;\"-gdwarf64\"

Total Discovered Tests: 33363
  Skipped          :     1 (0.00%)
  Unsupported      :   518 (1.55%)
  Passed           : 32273 (96.73%)
  Expectedly Failed:    31 (0.09%)
  Unresolved       :    12 (0.04%)
  Timed Out        :     1 (0.00%)
  Failed           :   527 (1.58%)
```
A lot of failures, and I had to kill some tests. Could be the way I'm passing the argument, but if so, I'd expect more failures in the baseline.

Running tests on your PR right now.

DavidSpickett commented:
DWARF64, with this PR branch:
```
Total Discovered Tests: 33363
  Skipped          :     1 (0.00%)
  Unsupported      :   518 (1.55%)
  Passed           : 32716 (98.06%)
  Expectedly Failed:    31 (0.09%)
  Unresolved       :     3 (0.01%)
  Failed           :    94 (0.28%)
```
And I did not need to kill any tests.

So thats...  433 fewer tests failing and 9 fewer tests unresolved, +443 passes.

A common error message I see is:
```
error: a.out Failed to extract range list table at offset 0x0000000000000014: .debug_rnglists table at offset 0x8 has too small length (0x4) to contain a complete header
```
And by searching that error message I found:
```
const std::optional<llvm::DWARFDebugRnglistTable> &
DWARFUnit::GetRnglistTable() {
  if (GetVersion() >= 5 && !m_rnglist_table_done) {
    m_rnglist_table_done = true;
    if (auto table_or_error =
            ParseListTableHeader<llvm::DWARFDebugRnglistTable>(
                GetRnglistData().GetAsLLVMDWARF(), m_ranges_base, DWARF32))
      m_rnglist_table = std::move(table_or_error.get());
    else
      GetSymbolFileDWARF().GetObjectFile()->GetModule()->ReportError(
          "Failed to extract range list table at offset {0:x16}: {1}",
          m_ranges_base, toString(table_or_error.takeError()).c_str());
  }
  return m_rnglist_table;
}
```
So this is your next thing to fix :)

DavidSpickett commented:
> Could use a test case.

But I too still want a targeted test case for this. Especially given that no one is going to be running a DWARF64 buildbot any time soon.

HemangGadhavi commented:
> I suggest you look in lldb's source for any uses of the DWARF32 symbol, I bet we've done this elsewhere too.

Yes I already found it and its in my list to fix, there are few places where `DWARF32` parsing but I was planing to do in separate PR. or you want me to do it here only ?

HemangGadhavi commented:
> This is the relevant part of the standard, correct?

Yes exactly that's what I am referring too. 

> I see also "7.4 32-Bit and 64-Bit DWARF Formats", is there anything else in there we need to handle?

I see there is one form `DW_FORM_strp_sup` which is not at all added in `LLDB` code but `LLVM` handles it, I am trying to generate the same, if so I can add this FORM as well.

DavidSpickett commented:
> Yes I already found it and its in my list to fix, there are few places where DWARF32 parsing but I was planing to do in separate PR. or you want me to do it here only ?

Separate PR please.

HemangGadhavi committed:
Added testcase

HemangGadhavi commented:
Hi @labath @DavidSpickett 
I have addressed your comments, and added separate test case to test DW_AT_str_offsets_base for DWARF5 & 64 format.
Please review once and give your inputs.

labath reviewed (approved):
Cool. Thanks.

DavidSpickett reviewed (approved):
LGTM.

HemangGadhavi committed:
changed LE to BE

HemangGadhavi commented on lldb/unittests/SymbolFile/DWARF/DWARF64UnitTest.cpp:
```diff
@@ -84,3 +84,100 @@ TEST(DWARF64UnitTest, DWARF64DebugInfoAndCU) {
+--- !ELF
+FileHeader:
+  Class:   ELFCLASS64
+  Data:    ELFDATA2LSB
```
Thanks Modified to Big-endian.
